Add "Admissions from supervision to prison over time" chart & new filters

nasaownsky opened the pull request:
## Description of the change

1. Added "Admissions from supervision to prison over time" chart 
2. Added DISTRICT filter to the `supervisionToPrisonOverTime` metric
3. Added MOST_SEVERE_VIOLATION and NUMBER_OF_VIOLATIONS filters to `supervisionToPrisonOverTime` metric

## Type of change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Configuration change (adjusts configuration to achieve some end related to functionality, development, performance, or security)

## Related issues

Closes #1450 
Closes #1452 
Closes #1453 

## Checklists

### Development

These boxes should be checked by the submitter prior to merging:

- [x] Manual testing against realistic data has been performed locally
- [ ] E2E have been run for Lantern changes ([Steps in the Readme](https://github.com/Recidiviz/pulse-dashboard#running-e2e-tests))

### Code review

These boxes should be checked by reviewers prior to merging:

- [x] This pull request has a descriptive title and information useful to a reviewer
- [x] This pull request has been moved out of a Draft state, has no "Work In Progress" label, and has assigned reviewers
- [x] Potential security implications or infrastructural changes have been considered, if relevant

nasaownsky commented:
Hey, @jovergaag! Looks like everything works fine, except new filters logic. Any ideas how to make filter separation by state? At first I was thinking about adding them to the `tenants.ts` file, but filters also can differ from metric to metric, and using them from `tenants.ts` file might be an inconvenient decision. Maybe use them from state copy files? But again this files is only for copy text. I would really appreciate any advices and implementation tips! Thanks.

jovergaag commented:
> 



> Hey, @jovergaag! Looks like everything works fine, except new filters logic. Any ideas how to make filter separation by state? At first I was thinking about adding them to the `tenants.ts` file, but filters also can differ from metric to metric, and using them from `tenants.ts` file might be an inconvenient decision. Maybe use them from state copy files? But again this files is only for copy text. I would really appreciate any advices and implementation tips! Thanks.

Good question. Maybe we can create a new object similar to `filterOption` that is called `enabledFilters`, probably also live in `src/core/utils`. It could be structured similarly to `filterOptions` with a key for each enabled state/tenant. Each state/tenant would have a key for each metric type. Similarly to `filterOptions` it would be accessible via the filters store. Does that seem like it will work?

jovergaag commented on src/core/models/SupervisionCountOverTimeMetric.ts:
```diff
@@ -55,6 +58,9 @@ export default class SupervisionCountOverTimeMetric extends PathwaysMetric<Super
           record.supervisionType === supervisionType &&
+          district.includes(record.district) &&
+          mostSevereViolation.includes(record.mostSevereViolation) &&
+          numberOfViolations.includes(record.numberOfViolations) &&
```
This metric will need to have the results grouped and summed, similar to lines 81-91 of `PopulationOverTimeMetric`. Currently, when more than one option of any given filter are selected the chart is not making sense. That is because the results are not summed properly.

jovergaag reviewed (commented):
Let me know how the `enabledFilters` work goes. I think what I suggested will work out, but feel free to play around with it.

nasaownsky commented:
Check my implementation out! This is how I see it, let me know your thoughts. Also, could you please try to debug that test error, I honestly tried, but had no success. And one more thing: I noticed, that `Supervision to Liberty` chart is broken in `US_ID`, because we changed `SupervisionCountOverTimeMetric`. So, will the files for it also be updated, or should I split metric into two instances?

jovergaag reviewed (changes requested):
This looks amazing! I pushed up a couple of commits with copy updates and fixes to those tests.

There are a few little things I commented on. The other remaining thing is the download is broken, something around creating the methodology file.

Don't worry about the US_ID Supervision to Liberty metric file - I have a task to update the metric file for US_ID before we land this.

jovergaag commented on src/core/PageSystem/PageSystem.tsx:
```diff
@@ -43,13 +43,8 @@ const PageSystem: React.FC = () => {
-    isLoading,
-    note,
-  } = metric;
+  const { filters, download, isLoading, note } = metric;
```
I'm getting an error on this line when I first open the page. After refreshing it goes away because the browser cache changes the timing. I think we have something going on with the `page` on line 61 of `MetricsStore`, it isn't properly updating to be "prison" before calculating the current metric based on the page/section. That might need a little work.

nasaownsky replied:
Could you try to do this please? I can't make it work for some reason

jovergaag commented:
One more small styling tweak, please decrease the margin here from 3.5rem to 2.5rem (for all section nav links including US_ID, not only this one):
![image (1)](https://user-images.githubusercontent.com/5872651/143324394-30fa9fbb-4bc7-4116-872f-50687c5fb3c3.png)

nasaownsky commented on src/core/content/page/us_tn.ts:
```diff
@@ -32,6 +32,10 @@ const content: StateSpecificPageCopy = {
     methodology:
       "These charts show people incarcerated in a TDOC prison for a felony or misdemeanor conviction. People in county jails are not included.",
   },
+  supervisionToPrison: {
+    methodology:
+      "Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris.",
+  },
```
The `download` wasn't work because you forgot to add methodology text for `supervisionToPrison` here

jovergaag replied:
Oops. We don't have the "default" methodology for this chart set in the copy doc. thank you!

nasaownsky commented on src/core/models/SupervisionCountOverTimeMetric.ts:
```diff
@@ -43,22 +48,45 @@ export default class SupervisionCountOverTimeMetric extends PathwaysMetric<Super
+      groupBy((d: SupervisionCountTimeSeriesRecord) => [d.year, d.month]),
+      values,
+      map((dataset) => ({
+        year: dataset[0].year,
+        month: dataset[0].month,
+        gender: dataset[0].gender,
+        supervisionType: dataset[0].supervisionType,
+        district: dataset[0].district,
+        mostSevereViolation: dataset[0].mostSevereViolation,
+        numberOfViolations: dataset[0].numberOfViolations,
```
Ok, I will think about how this can be made.

nasaownsky commented on src/core/PathwaysFilterBar/PathwaysFilterBar.tsx:
```diff
@@ -110,7 +110,12 @@ const PathwaysFilterBar: React.FC<{
               <CoreSelect
                 value={
-                  getFilterOptions(get(filters, filter.type), filter.options)[0]
+                  filter.type === FILTER_TYPES.TIME_PERIOD
+                    ? getFilterOption(get(filters, filter.type), filter.options)
+                    : getFilterOptions(
+                        get(filters, filter.type),
+                        filter.options
+                      )[0]
```
Also `timePeriod` wasn't correctly displayed, so I fixed this little issue.

jovergaag replied:
i have just gone through and done a fairly significant overhaul on the filters in my multi-select PR. All of this should be a little more clean moving forward. Thanks for fixing this here!

jovergaag reviewed (commented):
This looks great! I'm going to hold off approving/merging this until we can get the schema change in the backend to make sure that `main` is always deployable. That should be done within the next couple of days.

jovergaag reviewed (approved):
A ton of great work here. Thank you @nasaownsky
